modules/core/src/main/java/org/apache/synapse/message/store/impl/rabbitmq/RabbitMQProducer.java

```diff
@@ -36,147 +36,186 @@
-			if (exchangeName == null) {
-				channel.basicPublish("", queueName, builder.build(), byteForm);
-			} else {
-				channel.basicPublish(exchangeName, queueName, builder.build(), byteForm);
-			}
-		} catch (IOException e) {
-			throwable = e;
-			error = true;
-			isConnectionError = true;
-		} catch (Throwable t) {
-			throwable = t;
-			error = true;
-		} finally {
-			if (channel != null && channel.isOpen())
-				try {
-					channel.close();
-				} catch (TimeoutException e) {
-					logger.error("Error when closing connection,TimeoutException" + synCtx.getMessageID() + ". " , e);
-				} catch (IOException e) {
-					logger.error(
-							"Error when closing connection" + synCtx.getMessageID() + ". " + e);
-				}
-		}
-		if (error) {
-			String errorMsg = getId() + ". Ignored MessageID : " + synCtx.getMessageID()
-			                  + ". Could not store message to store ["
-			                  + store.getName() + "]. Error:" + throwable.getLocalizedMessage();
-			logger.error(errorMsg, throwable);
-			store.closeProducerConnection();
-			connection = null;
-			if (logger.isDebugEnabled()) {
-				logger.debug(getId() + ". Ignored MessageID : " + synCtx.getMessageID());
-			}
-			return false;
-		}
-		if (logger.isDebugEnabled()) {
-			logger.debug(getId() + ". Stored MessageID : " + synCtx.getMessageID());
-		}
-		store.enqueued();
-		return true;
-	}
-
-	/**
-	 * Used to close the channel opened in this object instance.
-	 * This should be called after the end of each call on storeMessage method
-	 * But instead of this, finally block is used in storeMethod to close the channels
-	 */
-	public boolean cleanup() {
-		return store.cleanup(null, false);
-	}
-
-	public boolean isInitialized() {
-		return isInitialized;
-	}
-
-	public String getId() {
-		return idString;
-	}
+    private static final Log log = LogFactory.getLog(RabbitMQProducer.class.getName());
+
+    private static final int DEFAULT_PRIORITY = 0;
+    private Connection connection;
+    private RabbitMQStore store;
+    private String routingKey;
+    private String exchangeName;
+    private boolean isInitialized = false;
+    private String idString; // ID of the MessageProducer
+
+    /**
+     * The RabbitMQ producer
+     *
+     * @param store the {@link RabbitMQStore} object
+     */
+    public RabbitMQProducer(RabbitMQStore store) {
+        if (store == null) {
+            log.error("Cannot initialize.");
+            return;
+        }
+        this.store = store;
+        isInitialized = true;
+    }
+
+    /**
+     * Store the given message into the queue and return whether the operation success or not
+     *
+     * @param synCtx Message to be saved.
+     * @return {@code true} if storing of the message is successful, {@code false} otherwise.
+     */
+    @Override
+    public boolean storeMessage(MessageContext synCtx) {
+        if (synCtx == null) {
+            return false;
+        }
+        if (connection == null) {
```

Comment:
Message store and processor is used for guaranteed delivery scenarios. If we just keep shut up upon connection is null, the flow that is trying to store the message doesn't get that guarantee. Isn't it?

Reply:
The store mediator invokes the fault sequence in such a case. The client has to retry based on the response sent in the fault sequence. The mediation engine doesn't suppose to retry internally.